src/vulkan-renderer/input/keyboard_mouse_data.cpp

```diff
@@ -1,75 +1,87 @@
 #include "inexor/vulkan-renderer/input/keyboard_mouse_data.hpp"
 
+#include <mutex>
+
 namespace inexor::vulkan_renderer::input {
 
 void KeyboardMouseInputData::press_key(const std::int32_t key) {
+    std::scoped_lock lock(m_input_mutex);
     m_pressed_keys[key] = true;
-    keyboard_updated = true;
+    m_keyboard_updated = true;
 }
 
 void KeyboardMouseInputData::release_key(const std::int32_t key) {
+    std::scoped_lock lock(m_input_mutex);
     m_pressed_keys[key] = false;
-    keyboard_updated = true;
+    m_keyboard_updated = true;
+}
+
+bool KeyboardMouseInputData::is_key_pressed(const std::int32_t key) const {
+    std::shared_lock lock(m_input_mutex);
+    return m_pressed_keys[key];
+}
+bool KeyboardMouseInputData::was_key_pressed_once(const std::int32_t key) {
+    std::scoped_lock lock(m_input_mutex);
+    if (!m_keyboard_updated) {
+        return false;
+    }
+    if (m_pressed_keys[key]) {
+        m_pressed_keys[key] = false;
+        return true;
+    }
+    return false;
 }
 
 void KeyboardMouseInputData::press_mouse_button(const std::int32_t button) {
+    std::scoped_lock lock(m_input_mutex);
     m_pressed_mouse_buttons[button] = true;
-    mouse_buttons_updated = true;
+    m_mouse_buttons_updated = true;
 }
 
 void KeyboardMouseInputData::release_mouse_button(const std::int32_t button) {
+    std::scoped_lock lock(m_input_mutex);
     m_pressed_mouse_buttons[button] = false;
-    mouse_buttons_updated = true;
+    m_mouse_buttons_updated = true;
 }
 
-void KeyboardMouseInputData::set_cursor_pos(const double cursor_pos_x, const double cursor_pos_y) {
-    m_current_cursor_pos[0] = static_cast<std::int64_t>(cursor_pos_x);
-    m_current_cursor_pos[1] = static_cast<std::int64_t>(cursor_pos_y);
+bool KeyboardMouseInputData::is_mouse_button_pressed(const std::int32_t button) const {
+    std::shared_lock lock(m_input_mutex);
+    return m_pressed_mouse_buttons[button];
 }
 
-std::array<std::int64_t, 2> KeyboardMouseInputData::get_cursor_pos() const {
-    return m_current_cursor_pos;
+bool KeyboardMouseInputData::was_mouse_button_pressed_once(const std::int32_t button) {
+    std::scoped_lock lock(m_input_mutex);
+    if (!m_mouse_buttons_updated) {
+        return false;
+    }
+    if (m_pressed_mouse_buttons[button]) {
+        m_pressed_mouse_buttons[button] = false;
+        return true;
+    }
+    return false;
```

Comment:
Check can be simplified.
```suggestion
    if (!m_mouse_buttons_updated || !m_pressed_mouse_buttons[button]) {
        return false;
    }
    m_pressed_mouse_buttons[button] = false;
    return true;
```

Reply:
Done.